File: salmon_king/sim/weather.py
from __future__ import annotations

from salmon_king.data.camps import CampDef
from salmon_king.sim.clock import doy
from salmon_king.sim.models import Weather
from salmon_king.sim.rng import Rng


def roll_weather(day, camp: CampDef, rng: Rng, force: Weather | None = None) -> Weather:
    if force:
        return force
    n = doy(day)
    # June–July advection fog; September gets mean
    fog_p = camp.fog + (0.18 if 152 <= n <= 212 else 0.0)
    will_p = camp.williwaw + (0.08 if n >= 244 else 0.0)
    wind = max(4.0, rng.gauss(14.0, 6.0))
    if n >= 244:
        wind += rng.uniform(2, 10)
    seas = max(0.5, wind / 6.0 + rng.uniform(-0.4, 1.2))
    if camp.district == "central":
        seas += 0.4  # Shelikof short steep
    fog = min(1.0, max(0.0, rng.gauss(fog_p, 0.15)))
    williwaw = rng.random() < will_p * 0.08
    if williwaw:
        wind = max(wind, rng.uniform(42, 68))
        seas = max(seas, rng.uniform(4.5, 8.0))
        precip = "williwaw"
        label = f"WILLIWAW {int(wind)} kt, {seas:.0f} ft"
    elif fog > 0.55:
        precip = "fog"
        label = f"advection fog, {int(wind)} kt"
    elif rng.random() < 0.45:
        precip = "rain"
        label = f"rain {int(wind)} kt, {seas:.0f} ft"
    else:
        precip = "overcast"
        label = f"overcast {int(wind)} kt, {seas:.0f} ft"
    return Weather(
        wind_kt=round(wind, 1),
        seas_ft=round(seas, 1),
        fog=round(fog, 2),
        precip=precip,
        williwaw=williwaw,
        label=label,
    )


def skiffs_grounded(weather: Weather) -> bool:
    return weather.williwaw or weather.seas_ft >= 7.5 or weather.wind_kt >= 45

File: src/sim/weather.rs
use crate::data::camps::CampDef;
use crate::sim::clock::{doy, GameDate};
use crate::sim::models::Weather;
use crate::sim::rng::Rng;

pub fn roll_weather(day: GameDate, camp: &CampDef, rng: &mut Rng) -> Weather {
    let n = doy(day);
    let fog_p = camp.fog + if (152..=212).contains(&n) { 0.18 } else { 0.0 };
    let will_p = camp.williwaw + if n >= 244 { 0.08 } else { 0.0 };
    let mut wind = rng.gauss(14.0, 6.0).max(4.0);
    if n >= 244 {
        wind += rng.uniform(2.0, 10.0);
    }
    let mut seas = (wind / 6.0 + rng.uniform(-0.4, 1.2)).max(0.5);
    if camp.district == "central" {
        seas += 0.4;
    }
    let fog = rng.gauss(fog_p, 0.15).clamp(0.0, 1.0);
    let williwaw = rng.random() < will_p * 0.08;
    let (precip, label) = if williwaw {
        wind = wind.max(rng.uniform(42.0, 68.0));
        seas = seas.max(rng.uniform(4.5, 8.0));
        (
            "williwaw".to_string(),
            format!("WILLIWAW {} kt, {:.0} ft", wind as i32, seas),
        )
    } else if fog > 0.55 {
        (
            "fog".to_string(),
            format!("advection fog, {} kt", wind as i32),
        )
    } else if rng.random() < 0.45 {
        (
            "rain".to_string(),
            format!("rain {} kt, {:.0} ft", wind as i32, seas),
        )
    } else {
        (
            "overcast".to_string(),
            format!("overcast {} kt, {:.0} ft", wind as i32, seas),
        )
    };
    Weather {
        wind_kt: (wind * 10.0).round() / 10.0,
        seas_ft: (seas * 10.0).round() / 10.0,
        fog: (fog * 100.0).round() / 100.0,
        precip,
        williwaw,
        label,
    }
}

pub fn skiffs_grounded(weather: &Weather) -> bool {
    weather.williwaw || weather.seas_ft >= 7.5 || weather.wind_kt >= 45.0
}
